util/coverage/coverage_off_target.py

```diff
@@ -0,0 +1,253 @@
+
+Typical usage:
+    util/coverage/coverage_off_target.py --print-text-report
+"""
+
+import logging as log
+import subprocess
+from enum import Enum
+from pathlib import Path
+from random import randint
+from pprint import pformat
+from typing import List
+
+import typer
+
+app = typer.Typer(pretty_exceptions_enable=False, add_completion=False)
+
+# Commands that this script uses
+BAZEL = "./bazelisk.sh"
+LLD = "ld.lld"
+LLVM_COV = "llvm-cov"
+LLVM_PROFDATA = "llvm-profdata"
+
+# Title of the HTML coverage report
+REPORT_TITLE = "OpenTitan Unit Test Coverage"
+
+
+class LogLevel(str, Enum):
+    none = "none"
+    info = "info"
+    debug = "debug"
+
+
+def run(*args):
+    """Run the given command in a subprocess.
+
+    Args:
+        args: Command (the first parameter) and arguments (remaining arguments).
+
+    Returns:
+        Stdout lines in a list. Empty lines are filtered out.
+    """
+    log.debug(f"command: {' '.join(args)}")
+    try:
+        res = subprocess.run(args,
+                             stdout=subprocess.PIPE,
+                             stderr=subprocess.PIPE,
+                             encoding="utf-8",
+                             check=True)
+    except subprocess.CalledProcessError as e:
+        log.error(f"stdout: {e.stdout if e.stdout else '(empty)'}")
+        log.error(f"stderr: {e.stderr if e.stderr else '(empty)'}")
+        raise
+    log.debug(f"stdout: {res.stdout if res.stdout else '(empty)'}")
+    log.debug(f"stderr: {res.stderr if res.stderr else '(empty)'}")
+    return [line for line in res.stdout.splitlines() if line]
+
+
+def gather_and_build_libs():
+    """Find all device libraries that can be built on the host build them with coverage
+    instrumentation.
+
+    Returns:
+        The list of device libraries that can be built on the host.
+    """
+    # Query for device libraries to be instrumented.
+    DEVICE_LIBS_INC = [
+        "//sw/device/silicon_creator/...",
+        "//sw/device/lib/...",
+    ]
+    DEVICE_LIBS_EXC = [
+        "//sw/device/lib/dif/...",
+        "//sw/device/lib/testing/... ",
+        "//sw/device/lib/runtime/...",
+        "//sw/device/lib/crypto/...",
+        "//sw/device/lib/arch/...",
+        "//sw/device/lib/ujson/...",
+        "//sw/device/lib/base:status",
+        "//sw/device/lib/base:mmio_on_device_do_not_use_directly",
+        "//sw/device/lib/base:mmio_on_host_do_not_use_directly",
+        "//sw/device/silicon_creator/rom_ext/...",
+        "//sw/device/silicon_creator/rom/keys/real:real",
+    ]
+    DEVICE_LIBS_QUERY = (f"kind(cc_library, ({' + '.join(DEVICE_LIBS_INC)}) "
+                         f"- ({' + '.join(DEVICE_LIBS_EXC)}))")
+
+    # Gather and build all device libs that can be built on host.
+    # We handle everyting in a single cquery instead of one cquery per test target since
+    # each invocation incurs an additional startup cost.
+    device_libs_all = run(BAZEL, "cquery", DEVICE_LIBS_QUERY,
```

Comment:
Is there a reason for choosing `cquery` over regular `query`?

I recently noticed in the Bazel docs that `cquery`'s output is nondeterministic. It apparently is influenced by previous builds. https://bazel.build/query/cquery#known-issues

Reply:
No, this can use the regular `query`, done.